fdbserver/include/fdbserver/TenantEntryCache.actor.h

```diff
@@ -0,0 +1,367 @@
+	T payload;
+};
+
+template <class T>
+using TenantEntryCachePayloadFunc =
+    std::function<TenantEntryCachePayload<T>(const TenantName&, const TenantMapEntry&, const TenantEntryCacheGen)>;
+
+// In-memory cache for TenantEntryMap objects. It supports three indices:
+// 1. Lookup by 'TenantId'
+// 2. Lookup by 'TenantPrefix'
+// 3. Lookup by 'TenantName'
+//
+// TODO:
+// ----
+// The cache allows user to construct the 'cached object' by supplying a callback to allow constructing the cached
+// object. The cache implements a periodic refresh mechanism, polling underlying database for updates (add/remove
+// tenants), in future we might want to implement database range-watch to monitor such updates
+
+template <class T>
+class TenantEntryCache : public ReferenceCounted<TenantEntryCache<T>>, NonCopyable {
+private:
+	UID uid;
+	Database db;
+	TenantEntryCacheGen curGen;
+	TenantEntryCachePayloadFunc<T> createPayloadFunc;
+
+	AsyncTrigger sweepTrigger;
+	Future<Void> sweeper;
+	Future<Void> refresher;
+	Map<int64_t, TenantEntryCachePayload<T>> mapByTenantId;
+	Map<TenantName, TenantEntryCachePayload<T>> mapByTenantName;
+
+	CounterCollection metrics;
+	Counter hits;
+	Counter misses;
+	Counter refreshByCacheInit;
+	Counter refreshByCacheMiss;
+	Counter numSweeps;
+	Counter numRefreshes;
+
+	ACTOR static Future<TenantNameEntryPairVec> getTenantList(Reference<ReadYourWritesTransaction> tr) {
+		tr->setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+		tr->setOption(FDBTransactionOptions::READ_LOCK_AWARE);
+
+		KeyBackedRangeResult<std::pair<TenantName, TenantMapEntry>> tenantList =
+		    wait(TenantMetadata::tenantMap().getRange(
+		        tr, Optional<TenantName>(), Optional<TenantName>(), CLIENT_KNOBS->MAX_TENANTS_PER_CLUSTER + 1));
+		ASSERT(tenantList.results.size() <= CLIENT_KNOBS->MAX_TENANTS_PER_CLUSTER && !tenantList.more);
+
+		TraceEvent(SevDebug, "TenantEntryCache.GetTenantList").detail("Count", tenantList.results.size());
+
+		return tenantList.results;
+	}
+
+	static void updateCacheRefreshMetrics(TenantEntryCache<T>* cache, TenantEntryCacheRefreshReason reason) {
+		if (reason == TenantEntryCacheRefreshReason::INIT) {
+			cache->refreshByCacheInit += 1;
+		} else if (reason == TenantEntryCacheRefreshReason::CACHE_MISS) {
+			cache->refreshByCacheMiss += 1;
+		}
+
+		cache->numRefreshes += 1;
+	}
+
+	ACTOR static Future<Void> refreshImpl(TenantEntryCache<T>* cache, TenantEntryCacheRefreshReason reason) {
+		TraceEvent(SevDebug, "TenantEntryCache.RefreshStart", cache->id()).detail("Reason", static_cast<int>(reason));
+
+		// Increment cache generation count
+		cache->incGen();
+		ASSERT(cache->getCurGen() < std::numeric_limits<uint64_t>::max());
+
+		state Reference<ReadYourWritesTransaction> tr = makeReference<ReadYourWritesTransaction>(cache->getDatabase());
+		loop {
+			try {
+				state TenantNameEntryPairVec tenantList = wait(getTenantList(tr));
+
+				for (auto& tenant : tenantList) {
+					cache->put(tenant);
+				}
```

Comment:
I see that this is using an asynchronous job to erase entries from the cache. For my understanding, can you elaborate on the motivation for this as opposed to clearing and replacing the cache when you read the tenant list?

Reply:
Cache currently is designed as a `read-through` cache; on map-lookup failure it does reload all cache entries; the behavior will be update to having a `keyrange-database` watch to handle add/remove/update entries of TenantPrefix subspace.

Reply:
Right, but what is the rationale for doing it this way if we are reading the entire list each time (or if we can get a list of specific updates via a watch mechanism)? Do the old generations have some benefit?

Reply:
Idea is not to actively cleanup entries, the foreground task just loads the latest data and then asynchronously old-gen data is cleaned up (classic mark/sweep algo). Are the concerns around oldGen key being accessed? If yes, the code can be (or rather should be) modified to return success **iff** keyGen == curGen.

Reply:
Updated the code to return success **iff** cached cell belongs to current active generation.

Reply:
My question is if we are already reading every entry and inserting them into the map, why not just clear the map and refill it? Then there won't be multiple generations present and there wouldn't be a need to do the sweep.

Reply:
Fair, I think memory consumption to collect all new entries is already done, so better to simplify the logic and replace the cache content in `refreshImpl` itself and not have a separate `sweep` phase (also given it is single process execution mode, so having a asyn sweep isn't impacting perf goal in anyways).
Thanks